packages/blocks/src/__internal__/utils/selection.ts

```diff
@@ -465,62 +468,64 @@ export function handleNativeRangeDragMove(
-}
-
-function isBlankAreaBeforeFirstBlock(startContainer: HTMLElement) {
-  if (!(startContainer instanceof HTMLElement)) return false;
-  return startContainer.className.includes('affine-frame-block-container');
-}
-
 export function isBlankArea(e: SelectionEvent) {
   const { cursor } = window.getComputedStyle(e.raw.target as Element);
   return cursor !== 'text';
 }
 
+// fix selection back to the nearest block
+// when user click on the edge of page (page mode) or frame (edgeless mode)
+function handleClickReTargeting(page: Page, e: SelectionEvent) {
+  const targetEl = getElementFromEventTarget(e.raw.target);
+  const block = targetEl?.closest(`[${BLOCK_ID_ATTR}]`) as {
+    model?: BaseBlockModel;
+    pageModel?: BaseBlockModel;
+  } | null;
+  const parentModel = block?.model || block?.pageModel;
+  if (!parentModel) return;
+
+  const shouldReTarget = matchFlavours(parentModel, [
+    'affine:frame',
+    'affine:page',
+  ]);
+  if (!shouldReTarget) return;
+
+  const { clientX, clientY } = e.raw;
+
+  const horizontalElement = getClosestEditor(clientY);
+  if (!horizontalElement) return;
+
+  const model = getModelByElement(horizontalElement);
+  const rect = horizontalElement.getBoundingClientRect();
+  if (matchFlavours(model, nonTextBlock) && clientY > rect.bottom) {
+    const parent = page.getParent(model);
+    assertExists(parent);
+    const id = page.addBlockByFlavour('affine:paragraph', {}, parent.id);
+    asyncFocusRichText(page, id);
+    return;
+  }
+
+  if (clientX < rect.left) {
+    const range = setStartRange(horizontalElement);
+    resetNativeSelection(range);
+  } else {
+    const range = setEndRange(horizontalElement);
+    resetNativeSelection(range);
+  }
+}
+
 export function handleNativeRangeClick(page: Page, e: SelectionEvent) {
+  // if not left click
+  if (e.button) return;
+
   const range = caretRangeFromPoint(e.raw.clientX, e.raw.clientY);
   const startContainer = range?.startContainer;
-  // if not left click
-  if (e.button) {
-    return;
-  }
   // click on rich text
   if (startContainer instanceof Node) {
     resetNativeSelection(range);
   }
 
-  if (!(startContainer instanceof HTMLElement)) return;
-
-  if (
-    isBlankAreaBetweenBlocks(startContainer) ||
-    isBlankAreaBeforeFirstBlock(startContainer)
-  ) {
-    focusRichTextByOffset(startContainer, e.raw.clientX);
-  } else if (isBlankAreaAfterLastBlock(startContainer)) {
-    const { root } = page;
-    const lastChild = root?.lastChild();
-    assertExists(lastChild);
-    if (matchFlavours(lastChild, ['affine:paragraph', 'affine:list'])) {
-      const block = getBlockElementByModel(lastChild);
-      if (!block) return;
-      focusRichTextByOffset(block, e.raw.clientX);
-    } else if (matchFlavours(lastChild, ['affine:code', 'affine:embed'])) {
-      page.addBlockByFlavour(
-        'affine:paragraph',
-        {
-          text: new page.Text(page, ''),
-        },
-        page.getParent(lastChild)
-      );
-    }
-  }
```

Comment:
It's interesting that you actually have no way to trigger this part 🤣 
so I merge the part of handling non-text block.

The new version should now looks cleaner.